### PR TITLE
feat: throw a meaningful error when missing package.json

### PR DESCRIPTION
**Pull request by chengcyber**
### Summary

Closes https://github.com/pnpm/pn/issues/14

### CODE REVIEW

**Comment by chengcyber** [comment]
I'm excited to see this project comes alive! Hope to contribute more later :)

File: src/error.rs [hunk 1]
```diff
@@ -15,6 +15,9 @@ pub enum PnError {
     UnexpectedTermination { command: String },
     #[display(fmt = "--workspace-root may only be used in a workspace")]
     NotInWorkspace,
+    /// No package manifest
```
Comment (KSXGitHub) [inline src/error.rs:18]:
```suggestion
    /// No package manifest.
```

**Review by KSXGitHub — APPROVED** [review]

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
Test is missing.

File: src/main.rs [hunk 2]
```diff
@@ -53,7 +54,13 @@ fn run() -> Result<(), MainError> {
...
-                .map_err(MainError::from_dyn)?
+                .map_err(|err| {
+                    if err.kind() == ErrorKind::NotFound {
+                        MainError::Pn(PnError::NoPkgManifest { dir: cwd.clone() })
+                    } else {
+                        MainError::from_dyn(err)
+                    }
+                })?
                 .pipe(serde_json::de::from_reader::<_, NodeManifest>)
                 .map_err(MainError::from_dyn)?;
```
Comment (zkochan) [inline src/main.rs:65]: probably worth moving to a new function

File: src/main.rs [hunk 3]
```diff
@@ -54,13 +54,7 @@ fn run() -> Result<(), MainError> {
...
-                .map_err(|err| {
-                    if err.kind() == ErrorKind::NotFound {
-                        MainError::Pn(PnError::NoPkgManifest { dir: cwd.clone() })
-                    } else {
-                        MainError::from_dyn(err)
-                    }
-                })?
+                .map_err(|err| handle_no_package_manifest_error(err, cwd.clone()))?
                 .pipe(serde_json::de::from_reader::<_, NodeManifest>)
                 .map_err(MainError::from_dyn)?;
```
Comment (zkochan) [inline src/main.rs]: I meant reading and parsing the file in a new function

Comment (chengcyber) [reply src/main.rs]: working on it.

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
Good morning!

File: src/main.rs [hunk 4]
```diff
@@ -160,3 +157,39 @@ fn pass_to_sub(command: String) -> Result<(), MainError> {
...
+                let parent_path = manifest_path
+                    .parent()
+                    .expect("Could not get parent directory")
+                    .to_path_buf();
+                MainError::Pn(PnError::NoPkgManifest { dir: parent_path })
+            }
+            _ => MainError::from_dyn(err),
+        })?
+        .pipe(serde_json::de::from_reader::<_, NodeManifest>)
+        .map_err(MainError::from_dyn)
```
Comment (zkochan) [inline src/main.rs:175]: I think we also need a custom error for when the JSON couldn't be parsed.

File: src/main.rs [hunk 5]
```diff
@@ -160,3 +157,78 @@ fn pass_to_sub(command: String) -> Result<(), MainError> {
...
+    eprintln!("Read package manifest error:\n{read_package_manifest_error_display}\n");
+
+    let expected_display = format!(
+        "{:?}",
+        MainError::Pn(PnError::ParseJSONError {
+            file: package_json_path,
+            message: String::from("trailing comma at line 1 column 41")
+        })
+    );
+    assert_eq!(read_package_manifest_error_display, expected_display);
```
Comment (KSXGitHub) [inline src/main.rs:233]: The variable reads `_display` but the formatter is `{:?}` (debug). I am confused. What are you trying to do here? Are you just trying to compare 2 error objects without deriving `PatrialEq`? Or are you just trying to compare 2 error messages?

If it's the latter, then you should just call `.to_string()` for `read_package_manifest_error_display` and `format!("... {package_json_path:?} ...")` for `expected_display`. You should also rename these variables to `received_message` and `expected_message` respectively.

If, however, what you are trying to do is to test if the returning error is of `ParseJsonError`, just write `assert!(matches!(received_error, PnError::ParseJsonError { .. }))` (remember to log the error before doing the assertion).

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
I will resolve the first two of my own reviews myself, but you need to make decisions on the third one.

File: src/main.rs [hunk 6]
```diff
@@ -160,3 +157,77 @@ fn pass_to_sub(command: String) -> Result<(), MainError> {
...
+    eprintln!("Read package manifest error:\n{read_package_manifest_error_display}\n");
+
+    let expected_display = format!(
+        "{:?}",
+        MainError::Pn(PnError::ParseJsonError {
+            file: package_json_path,
+            message: String::from("trailing comma at line 1 column 41")
+        })
+    );
+    assert_eq!(read_package_manifest_error_display, expected_display);
```
Comment (KSXGitHub) [inline src/main.rs:232]: The variable reads `_display` but the formatter is `{:?}` (debug). I am confused. What are you trying to do here? Are you just trying to compare 2 error objects without deriving `PatrialEq`? Or are you just trying to compare 2 error messages?

If it's the latter, then you should just call `.to_string()` for `read_package_manifest_error_display` and `format!("... {package_json_path:?} ...")` for `expected_display`. You should also rename these variables to `received_message` and `expected_message` respectively.

If, however, what you are trying to do is to test if the returning error is of `ParseJsonError`, just write `assert!(matches!(received_error, PnError::ParseJsonError { .. }))` (remember to log the error before doing the assertion).

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
Alright, I've resolved the first 2 of my own reviews. The last one requires your conscious decision.

File: src/main.rs (hunk 6 above)
Comment (chengcyber) [reply src/main.rs:232]: I am in the airport right now. Will work on it later after reaching my laptop. Maybe this night or tmrw

**Comment by chengcyber** [comment]
Update: Test updated:

1. test the error type by "matches" PnError::ParseJsonError
2. test the error message by comparing display message

File: src/error.rs [hunk 7]
```diff
@@ -20,6 +20,14 @@ pub enum PnError {
     #[display(fmt = "--workspace-root may only be used in a workspace")]
     NotInWorkspace,
 
+    /// No package manifest.
+    #[display(fmt = "File not found: {file:?}")]
+    NoPkgManifest { file: PathBuf },
+
+    /// Parse JSON error.
+    #[display(fmt = "Parse JSON file {file:?} error: {message}")]
```
Comment (KSXGitHub) [inline src/error.rs:28]:
```suggestion
    #[display(fmt = "Failed to parse {file:?}: {message}")]
```

File: src/main.rs [hunk 8]
```diff
@@ -153,6 +150,80 @@ fn pass_to_sub(command: String) -> Result<(), MainError> {
...
+        MainError::Pn(PnError::ParseJsonError {
+            file: _,
+            message: _
+        })
+    ));
+
+    let received_message = received_error.to_string();
+    eprintln!("Read package manifest error:\n{received_message}\n");
+    let expected_message =
+        format!("Parse JSON file {package_json_path:?} error: trailing comma at line 1 column 41",);
```
Comment (KSXGitHub) [inline src/main.rs:223]:
```suggestion
        format!("Failed to parse {package_json_path:?}: trailing comma at line 1 column 41",);
```

File: src/main.rs [hunk 9]
```diff
@@ -153,6 +150,80 @@ fn pass_to_sub(command: String) -> Result<(), MainError> {
...
+    .unwrap();
+
+    let received_error = read_package_manifest(&package_json_path).unwrap_err();
+    assert!(matches!(
+        received_error,
+        MainError::Pn(PnError::ParseJsonError {
+            file: _,
+            message: _
+        })
+    ));
```
Comment (KSXGitHub) [inline src/main.rs:218]:
```suggestion
    dbg!(&received_error);
    assert!(matches!(
        received_error,
        MainError::Pn(PnError::ParseJsonError {
            file: _,
            message: _
        })
    ));
```

**Review by KSXGitHub — APPROVED** [review]

File: src/main.rs [hunk 10]
```diff
@@ -153,6 +150,81 @@ fn pass_to_sub(command: String) -> Result<(), MainError> {
...
+
+    let received_message = received_error.to_string();
+    eprintln!("Read package manifest error:\n{received_message}\n");
```
Comment (KSXGitHub) [inline src/main.rs:222]:
```suggestion
    eprintln!("MESSAGE:\n{received_message}\n");
```

**Review by zkochan — APPROVED** [review]

**Pull request merged**